### DISCUSSION TITLE
Fully Featured Web App

### DISCUSSION THREAD

**Discussion by EugeneAksyonov** [varabyte/kobweb#479] [Q&A] [answered]
Hello Dear Kobweb Developers!
I am researching at the moment what kind of modern Kotlin frameworks I ca n use in WebApp development. I have been mocking my sample app under Kotlin Wasm, KotlinJS+Wrappers(React), KVision, and now the turn of Kobweb :). I do constructing my Data+Domain+Presentation levels architecture expecting that I need at least essential features

- theme switching (ColorMode)
- routing(rememberPageContext())
- **localization**
- **global app state**(in memory) management and binding to composed elements for auto rendering purpose

I did not mention at wide scope of required widgets like Forms, Data Tables, Carousels, Badges, Cards, Charts, Spinners, etc. yet at all because I could construct them myself with the time as a plan B solution.
Ideally I want to use MVVM or MVI and DI(Koin) same way as it works for Mobile apps.
So far I see that Themes and Routing are implemented and I can use it, but that is all. Btw, my respect to MD implementation, what the  great feature! As far as I see environment is perfectly fine for "Single Page Sites" like a landing pages.

So **my question is** - am I over expecting from KobWeb wide functionality that I need or it is a long way mentioned on the roadmap somewhere, Is there any ports of redux like storages and gettext libs for KobWeb, probably?

**Comment by EchoEllet** [comment]
For localization, you can use Kotlin multiplatform libraries like [moko-resources](https://github.com/icerockdev/moko-resources) or build your own solution as I did in this [repo](https://github.com/freshplatform/alrayada-web/tree/main/site/src/jsMain/kotlin/net/freshplatform/alrayada_web/core/services/localization)

Build your own if it's simple and you know what you want from localizations

**Reply by EugeneAksyonov** [reply]
Thanks @ellet0, how did I missed that moko-resources can support Kotlin-JS already. Definitely I will try it. I did write simple localization by myself for mock-up purposes(yours is more professional), but I was lazy to implement plural forms and formatting which would be nice to have for the middle+ size project that I have in mind. Thanks for the repo link, it might help me to feel the KobWeb philosophy.

**Reply by EchoEllet** [reply]
professional? please, mine is unnecessarily complex with no advantages :)

it's quite simple actually, doesn't support any complex features and since it's a very custom solution (with strings in Kotlin code) it will not support services like [crowdin](https://crowdin.com/) which is important for big projects, keep that in mind so you don't have to migrate the translations all over the project later

But this project was quite good and I enjoyed translating the project, I would rather not get my hands dirty with XML files and navigating through files and building all the time

**Reply by EugeneAksyonov** [reply]
I like KISS too :-)
Any hints about App State store and safe update?

**Reply by EchoEllet** [reply]
Why don't you use Kotlin flows?

**Reply by EugeneAksyonov** [reply]
Yeah, Flows are the option too. I was avoiding to use it for such purpose before but as far as I have tested CompositionLocal and flows today in my mock-up it would satisfy my needs, I believe.

**Comment by bitspittle** [comment, marked as answer]
I'm glad @ellet0 stepped in to discuss localization. Kobweb does not have an opinionated solution about it yet. Honestly, I was surprised when I first investigated it how little clear guidance there is for localization in webdev (contrasted with, say, Android, where localization is one of the things they got right)

Note that https://fluen.se/ is a site built using Kobweb, and they have a rich localization experience. So it's possible, but unfortunately left up to the reader at this point.

I'm not sure exactly what you're looking for with Global App State, so maybe a specific example could help. At this point, I assume you can just use Compose for that sort of stuff. You can declare a top level mutable state somewhere for example, or better yet something like a [CompositionLocal](https://developer.android.com/jetpack/compose/compositionlocal).

Kobweb is still pretty early so the ecosystem is still somewhat bare, unfortunately! However, it is Kotlin/JS so for widgets you can import existing JS libraries in a pinch. One of Kobweb's users is managing and maintaining https://github.com/stevdza-san/KotlinBootstrap which you may find interesting as well.

**Reply by EugeneAksyonov** [reply]
Thanks for the suggestions and Stefan's project link, @bitspittle! Under "Global App State" I mean to have a single point of truth object that can mutate in a unidirectional data flow manner. CompositionLocal worth to try certainly.

**Reply by EchoEllet** [reply]
I think what he meant is to have one state that he can use all across the application and when updating it from one place it will be updated to all places, it's like the color mode in KobWeb

**Reply by EugeneAksyonov** [reply]
@ellet0, yes correct, I thought it was clear, but right ColorMode in KobWeb designed with CompositionLocal to 
 propagate theme over the composes down. I did the same way for my app State and so far so good.

**Reply by EugeneAksyonov** [reply]
@bitspittle it seems you are working on support of Compose 1.6.0+ versions where new Resources management was implemented. It supports now Localization in the MoKo project style. If so, the question about site Localization should be resolved by itself soon, yet there are lot of issues and bugs.
